File: packages/ruby/lib/opensms/resources/sender_ids.rb
# frozen_string_literal: true

require_relative "base"

module Opensms
  module Resources
    # The +sender_ids+ resource: registered sender IDs, availability checks,
    # fee quotes, documents and drafts. Accessed as +client.sender_ids+.
    class SenderIds < Base
      CREATE_FIELDS = %i[value kind countries use_case sample_message documents draft_id draft_version quote_id].freeze
      UPDATE_FIELDS = %i[use_case countries documents sample_message].freeze
      DRAFT_FIELDS = %i[source value kind countries use_case sample_message documents].freeze
      DRAFT_UPDATE_FIELDS = %i[version value kind countries use_case sample_message documents].freeze

      # GET /v1/sender-ids -> Page<SenderId>.
      def list(params = nil, **kwargs)
        http_page("/v1/sender-ids", Models.build(merge(params, kwargs), %i[limit cursor]))
      end

      # GET /v1/sender-ids/{id} -> SenderId.
      def get(id)
        http_get(path("/v1/sender-ids", id!(id)))
      end

      # Register a sender ID. May charge fees, so it is never auto-retried.
      # POST /v1/sender-ids -> 201 SenderId.
      #
      # @param params [Hash] :value, :kind, :countries, :documents (required),
      #   :use_case, :sample_message, :draft_id, :draft_version, :quote_id
      def create(params = nil, **kwargs)
        body = Models.build(merge(params, kwargs), CREATE_FIELDS, required: %i[value kind countries documents])
        @transport.request(:post, "/v1/sender-ids", body: body)
      end

      # Amend a sender ID. PATCH /v1/sender-ids/{id} -> SenderId.
      # @param params [Hash] :use_case, :countries, :documents (required), :sample_message
      def update(id, params = nil, **kwargs)
        body = Models.build(merge(params, kwargs), UPDATE_FIELDS, required: %i[use_case countries documents])
        @transport.request(:patch, path("/v1/sender-ids", id!(id)), body: body)
      end

      # DELETE /v1/sender-ids/{id} -> 204.
      # @return [nil]
      def delete(id)
        http_delete(path("/v1/sender-ids", id!(id)))
      end

      # GET /v1/sender-ids/check -> { valid:, available:, reserved:, reason: }.
      # @param params [Hash] :value (required), :country
      def check(params = nil, **kwargs)
        http_get("/v1/sender-ids/check", Models.build(merge(params, kwargs), %i[value country], required: %i[value]))
      end

      # Registration fee quote. GET /v1/sender-ids/quote?countries=KE,NG.
      # @param params [Hash] :countries (Array or comma-separated String)
      def quote(params = nil, **kwargs)
        http_get("/v1/sender-ids/quote", Models.build(merge(params, kwargs), %i[countries], required: %i[countries]))
      end

      # Uploaded registration documents. GET /v1/sender-documents.
      # @return [Array<Hash>] the +items+ of the response (no cursor)
      def list_documents
        body = http_get("/v1/sender-documents")
        body.is_a?(Hash) ? (body[:items] || []) : body
      end

      # GET /v1/sender-id-drafts -> Page<Draft>.
      def list_drafts(params = nil, **kwargs)
        http_page("/v1/sender-id-drafts", Models.build(merge(params, kwargs), %i[limit cursor]))
      end

      # POST /v1/sender-id-drafts -> 201 Draft. Not auto-retried.
      def create_draft(params = nil, **kwargs)
        @transport.request(:post, "/v1/sender-id-drafts", body: Models.build(merge(params, kwargs), DRAFT_FIELDS))
      end

      # GET /v1/sender-id-drafts/{id} -> Draft.
      def get_draft(id)
        http_get(path("/v1/sender-id-drafts", id!(id)))
      end

      # PATCH /v1/sender-id-drafts/{id} -> Draft. Requires the current
      # +version+ (optimistic lock; 409 on mismatch).
      def update_draft(id, params = nil, **kwargs)
        body = Models.build(merge(params, kwargs), DRAFT_UPDATE_FIELDS, required: %i[version])
        @transport.request(:patch, path("/v1/sender-id-drafts", id!(id)), body: body)
      end

      # DELETE /v1/sender-id-drafts/{id} -> 204.
      # @return [nil]
      def delete_draft(id)
        http_delete(path("/v1/sender-id-drafts", id!(id)))
      end
    end
  end
end

File: packages/python/src/opensms/resources/sender_ids.py
"""The ``sender_ids`` resource (sender IDs, drafts and the document list)."""

from __future__ import annotations

from typing import Any, List, Optional

from .._serialize import path_id, prune
from ..models import Page, SenderDocument, SenderId, SenderIdCheck, SenderIdDraft, SenderIdQuote
from ._base import Resource


class SenderIds(Resource):
    """Alphanumeric and numeric sender IDs. Accessed as ``client.sender_ids``."""

    def list(self, *, limit: Optional[int] = None, cursor: Optional[str] = None) -> Page[SenderId]:
        """List sender IDs."""
        return self._page(self._t.request("GET", "/v1/sender-ids", query={"limit": limit, "cursor": cursor}))

    def get(self, id: str) -> SenderId:
        """Fetch a sender ID with its registrations."""
        return self._t.request("GET", f"/v1/sender-ids/{path_id(id)}")

    def create(
        self,
        *,
        value: str,
        kind: str,
        countries: List[str],
        documents: List[Any],
        use_case: Optional[str] = None,
        sample_message: Optional[str] = None,
        draft_id: Optional[str] = None,
        draft_version: Optional[int] = None,
        quote_id: Optional[str] = None,
    ) -> SenderId:
        """Register a sender ID. May charge registration fees (see
        :meth:`quote`), so it is never retried automatically."""
        body = prune(
            {
                "value": value,
                "kind": kind,
                "countries": list(countries),
                "documents": list(documents),
                "use_case": use_case,
                "sample_message": sample_message,
                "draft_id": draft_id,
                "draft_version": draft_version,
                "quote_id": quote_id,
            }
        )
        return self._t.request("POST", "/v1/sender-ids", json_body=body)

    def update(
        self,
        id: str,
        *,
        use_case: str,
        countries: List[str],
        documents: List[Any],
        sample_message: Optional[str] = None,
    ) -> SenderId:
        """Amend a sender ID application."""
        body = prune(
            {
                "use_case": use_case,
                "countries": list(countries),
                "documents": list(documents),
                "sample_message": sample_message,
            }
        )
        return self._t.request("PATCH", f"/v1/sender-ids/{path_id(id)}", json_body=body)

    def delete(self, id: str) -> None:
        """Delete a sender ID."""
        self._t.request("DELETE", f"/v1/sender-ids/{path_id(id)}")

    def check(self, *, value: str, country: Optional[str] = None) -> SenderIdCheck:
        """Check whether a value is valid, available and not reserved."""
        return self._t.request("GET", "/v1/sender-ids/check", query={"value": value, "country": country})

    def quote(self, *, countries: List[str]) -> SenderIdQuote:
        """Quote registration fees for the given countries."""
        return self._t.request("GET", "/v1/sender-ids/quote", query={"countries": list(countries)})

    def list_documents(self) -> List[SenderDocument]:
        """List uploaded sender documents (upload itself is console only).
        Returns the ``items`` list; there is no cursor."""
        data = self._t.request("GET", "/v1/sender-documents")
        if isinstance(data, dict):
            return list(data.get("items") or [])
        return data or []

    def list_drafts(self, *, limit: Optional[int] = None, cursor: Optional[str] = None) -> Page[SenderIdDraft]:
        """List application drafts."""
        return self._page(
            self._t.request("GET", "/v1/sender-id-drafts", query={"limit": limit, "cursor": cursor})
        )

    def create_draft(
        self,
        *,
        source: Optional[str] = None,
        value: Optional[str] = None,
        kind: Optional[str] = None,
        countries: Optional[List[str]] = None,
        use_case: Optional[str] = None,
        sample_message: Optional[str] = None,
        documents: Optional[List[str]] = None,
    ) -> SenderIdDraft:
        """Start an application draft (``source`` ``application`` or
        ``onboarding``)."""
        body = prune(
            {
                "source": source,
                "value": value,
                "kind": kind,
                "countries": countries,
                "use_case": use_case,
                "sample_message": sample_message,
                "documents": documents,
            }
        )
        return self._t.request("POST", "/v1/sender-id-drafts", json_body=body)

    def get_draft(self, id: str) -> SenderIdDraft:
        """Fetch a draft."""
        return self._t.request("GET", f"/v1/sender-id-drafts/{path_id(id)}")

    def update_draft(
        self,
        id: str,
        *,
        version: int,
        source: Optional[str] = None,
        value: Optional[str] = None,
        kind: Optional[str] = None,
        countries: Optional[List[str]] = None,
        use_case: Optional[str] = None,
        sample_message: Optional[str] = None,
        documents: Optional[List[str]] = None,
    ) -> SenderIdDraft:
        """Update a draft. ``version`` must be the current version
        (optimistic lock, ``409`` on mismatch)."""
        body = prune(
            {
                "version": version,
                "source": source,
                "value": value,
                "kind": kind,
                "countries": countries,
                "use_case": use_case,
                "sample_message": sample_message,
                "documents": documents,
            }
        )
        return self._t.request("PATCH", f"/v1/sender-id-drafts/{path_id(id)}", json_body=body)

    def delete_draft(self, id: str) -> None:
        """Delete a draft."""
        self._t.request("DELETE", f"/v1/sender-id-drafts/{path_id(id)}")
